allowsFirstPartyForCookies MESSAGE_CHECK is causing WebContent process termination

charliewolfe opened the pull request:
#### 3781d0c8c8a6e6516c2c6d8d121299b05bd0eb11
<pre>
allowsFirstPartyForCookies MESSAGE_CHECK is causing WebContent process termination
<a href="https://bugs.webkit.org/show_bug.cgi?id=257872">https://bugs.webkit.org/show_bug.cgi?id=257872</a>
rdar://108747218

Reviewed by Alex Christensen.

Add logging in the web process which triggers before we fail this message check in the network process.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::addAllowedFirstPartyForCookies):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::handleProvisionalLoadFailureFromContentFilter):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::addAllowedFirstPartyForCookies):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::launchSharedWorker):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addAllowedFirstPartyForCookies):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::allowsFirstPartyForCookies):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::addAllowedFirstPartyForCookies):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadAlternateHTML):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::addAllowedFirstPartyForCookies):
(WebKit::WebProcess::allowsFirstPartyForCookies):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/265230@main">https://commits.webkit.org/265230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d720461cdc91b89b819bd5ed1196c548edb9008d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10275 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10778 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11922 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9891 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10464 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10433 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12308 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8473 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16586 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9572 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12706 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9909 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8038 "5 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9068 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2476 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13318 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9747 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -529,6 +529,11 @@ void WebLoaderStrategy::scheduleLoadFromNetworkProcess(ResourceLoader& resourceL
         existingNetworkResourceLoadIdentifierToResume = std::exchange(m_existingNetworkResourceLoadIdentifierToResume, std::nullopt);
     WEBLOADERSTRATEGY_RELEASE_LOG("scheduleLoad: Resource is being scheduled with the NetworkProcess (priority=%d, existingNetworkResourceLoadIdentifierToResume=%" PRIu64 ")", static_cast<int>(resourceLoader.request().priority()), valueOrDefault(existingNetworkResourceLoadIdentifierToResume).toUInt64());
+
+    auto sendResult = WebProcess::singleton().ensureNetworkProcessConnection().connection().sendSync(Messages::NetworkConnectionToWebProcess::AllowsFirstPartyForCookies(RegistrableDomain { loadParameters.request.firstPartyForCookies() }), 0);
```
Absolutely not.  Introducing a synchronous message send here would kill performance and cause hangs whenever anything is loaded.

charliewolfe replied:
Ok. I'll change this to keep track of the allowedFirstPartyForCookies in the web process so we don't need this synchronous IPC call.

achristensen07 commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
```diff
@@ -111,6 +111,8 @@ messages -> NetworkProcessProxy LegacyReceiver {
     ReportNetworkIssue(WebKit::WebPageProxyIdentifier webPageProxyIdentifier, URL requestURL)
 #endif
+
+    AddAllowedFirstPartyForCookies(WebCore::ProcessIdentifier webProcessIdentifier, WebCore::RegistrableDomain firstPartyDomain)
```
This should probably go through the NetworkConnectionToWebProcess instead of through the UI process.

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2319,6 +2319,32 @@ void WebProcess::setNetworkProcessConnectionID(IPC::Connection::UniqueID uniqueI
+
+bool WebProcess::allowsFirstPartyForCookies(const URL& firstParty)
+{
+    if (firstParty.isAboutBlank())
```
Can this duplicate code be shared somehow?

achristensen07 commented on Source/WebKit/Shared/AuxiliaryProcess.h:
```diff
@@ -30,6 +30,7 @@
 #include "MessageSender.h"
 #include "SandboxExtension.h"
 #include <WebCore/ProcessIdentifier.h>
+#include <WebCore/RegistrableDomain.h>
```
This can be a forward declaration instead of another included file.

achristensen07 commented on Source/WebKit/Shared/AuxiliaryProcess.h:
```diff
@@ -161,6 +162,10 @@ class AuxiliaryProcess : public IPC::Connection::Client, public IPC::MessageSend
     void didClose(IPC::Connection&) override;
 
+    bool allowsFirstPartyForCookies(const URL&);
+
```
nit: I'd remove this space.

achristensen07 commented on Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
```diff
@@ -30,6 +30,7 @@
 #include <JavaScriptCore/ConsoleTypes.h>
 #include <WebCore/MessagePortChannelProvider.h>
 #include <WebCore/RTCDataChannelIdentifier.h>
+#include <WebCore/RegistrableDomain.h>
```
Ditto.  Forward declaration instead of adding a #include

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -781,6 +784,8 @@ class WebProcess : public AuxiliaryProcess
     bool m_hadMainFrameMainResourcePrivateRelayed { false };
     bool m_imageAnimationEnabled { true };
+
+    HashSet<WebCore::RegistrableDomain> m_allowedFirstPartyForCookies;
```
nit: plural parties.

achristensen07 commented on Source/WebKit/Shared/AuxiliaryProcess.cpp:
```diff
@@ -250,4 +250,33 @@ void AuxiliaryProcess::didReceiveMemoryPressureEvent(bool isCritical)
+    if (firstParty.isNull())
+        return true; // FIXME: This shouldn't be allowed.
+
+    return false;
```
This isn't really "false" - this value should indicate that another check is needed.  I think it would be better if this function required a HashSet and passed it to the other AuxiliaryProcess::allowsFirstPartyForCookies instead of returning a misleading value here.

charliewolfe committed:
allowsFirstPartyForCookies MESSAGE_CHECK is causing WebContent process termination
https://bugs.webkit.org/show_bug.cgi?id=257872
rdar://108747218

Reviewed by Alex Christensen.

Add logging in the web process which triggers before we fail this message check in the network process.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::addAllowedFirstPartyForCookies):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::handleProvisionalLoadFailureFromContentFilter):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::addAllowedFirstPartyForCookies):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::launchSharedWorker):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addAllowedFirstPartyForCookies):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::allowsFirstPartyForCookies):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::addAllowedFirstPartyForCookies):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadAlternateHTML):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::addAllowedFirstPartyForCookies):
(WebKit::WebProcess::allowsFirstPartyForCookies):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/265230@main
